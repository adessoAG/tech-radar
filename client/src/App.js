import React, { useState, useEffect } from "react";

// SERVICES
import commentService from './services/commentService';

function App() {
    const [comments, setcomments] = useState(null);

    useEffect(() => {
        if(!comments) {
            getComments();
        }
    });

    const getComments = async () => {
        let res = await commentService.getByRadarType('microsoft');
        setcomments(res);
    };

    const getCommentsDB = () => {
        return comments;
    }

    const renderComments = comment => {
        return (
            <li key={comment._id} className="list__item comment">
                <h3 className="comment__name">{comment.autor}</h3>
                <p className="comment__description">{comment.zeit}</p>
                <p className="comment__description">{comment.text}</p>
                <p className="comment__description">{comment.meinung}</p>
                <p className="comment__description">{comment.technologie}</p>
                <p className="comment__description">{comment.radar}</p>
            </li>
        );
    };

    return (
        <div className="App">
            <ul className="list">
                {(comments && comments.length > 0) ? (
                    comments.map(comment => renderComments(comment))
                ) : (
                    <p>No comments found!</p>
                )}
            </ul>
        </div>
    );
}

export default App;
