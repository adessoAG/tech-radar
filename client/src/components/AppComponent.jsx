import React from "react";
import RadarDataService from "./RadarDataService";
import FAQ from "./FAQ";
import HotTopics from "./HotTopics";

import javaJSON from "./java-radar.json";
import jsJSON from "./javascript-radar.json";
import msJSON from "./microsoft-radar.json";

import Tabs from "@material-ui/core/Tabs";
import Tab from "@material-ui/core/Tab";
import Button from "@material-ui/core/Button"
import AppBar from "@material-ui/core/AppBar";
import CardHeader from '@material-ui/core/CardHeader';
import Card from '@material-ui/core/Card';

import {Switch, Route, Link, BrowserRouter as Router} from "react-router-dom";

import logo from "../static/img/adesso.svg";

import "../static/css/styles.scss";
import "../static/css/desctop.scss";
import "../static/css/mobile.scss";
import ProtectedRoute  from './ProtectedRoute';


export default class AppComponent extends React.Component {

    constructor(props) {
        super(props);
        this.state = {
            isLoading: true,
            value: 0
        };
    }

    tabs;
    componentDidMount = () => {
        this.initializeTab();
    };

    //Based on the url, set the state ""value", which is needed for tabs, to a specific value
    initializeTab = () => {
        if (window.location.href.indexOf("/") > -1) this.setState({value: 0});
        if (window.location.href.indexOf("/java") > -1) this.setState({value: 0});
        if (window.location.href.indexOf("/microsoft") > -1)
            this.setState({value: 1});
        if (window.location.href.indexOf("/javascript") > -1)
            this.setState({value: 2});
        if (window.location.href.indexOf("/faq") > -1) this.setState({value: 3});
        if (window.location.href.indexOf("/hottopics") > -1) this.setState({value: 4});
    };

    handleChange = (event, index) => {
        this.setState({
            value: index
        });
    };


    render() {
        return (
            <Router>
                <div id="techleiste">
                    <AppBar
                        className="tech-appbar"
                        title="Techradar"
                        position="static"
                        color="default"
                    >
                        <Tabs
                            className="tech-tabs"
                            centered
                            value={this.state.value}
                            onChange={this.handleChange}
                            textColor="primary"
                        >
                            <Tab
                                className="tech-tab"
                                label={
                                    <img
                                        src={logo}
                                        className="tech-tab-img"
                                        alt="adesso's company logo"
                                    />
                                }
                                component={Link}
                                to="/java"
                                value={0}
                            />

                            <Tab
                                className="tech-tab"
                                id="javaTab"
                                label="Java"
                                component={Link}
                                to="/java"
                                value={0}
                            />

                            <Tab
                                className="tech-tab"
                                id="microTab"
                                label="Microsoft"
                                component={Link}
                                to="/microsoft"
                                value={1}
                            />

                            <Tab
                                className="tech-tab"
                                id="jsTab"
                                label="JavaScript"
                                component={Link}
                                to="/javascript"
                                value={2}
                            />

                            <Tab
                                className="tech-tab"
                                id="hottopicTab"
                                label="Hot Topics"
                                component={Link}
                                to="/hottopics"
                                value={4}
                            />

                            <Tab
                                className="tech-tab"
                                id="faq"
                                label="FAQ"
                                component={Link}
                                to="/faq"
                                value={3}
                            />
                            <Button variant="contained" id="log out">
                                Log out
                            </Button>
                        </Tabs>
                    </AppBar>
                    <Card>
                        <CardHeader
                            title= {'Herzlich Willkommen , '+localStorage.getItem('name')+'. Sie sind erfolgreich eingeloggt!'}>

                        </CardHeader>

                    </Card>
                </div>

                <Switch>
                    <Route
                        exact
                        path="/"
                        component={() => <RadarDataService data={javaJSON}/>}
                    />
                    <ProtectedRoute
                        path="/java"
                        component={() => <RadarDataService data={javaJSON}/>}
                    />
                    <ProtectedRoute
                        path="/microsoft"
                        component={() => <RadarDataService data={msJSON}/>}
                    />
                    <ProtectedRoute
                        path="/javascript"
                        component={() => <RadarDataService data={jsJSON}/>}
                    />
                    <ProtectedRoute
                        path="/hottopics"
                        component={HotTopics}
                    />
                    <ProtectedRoute path="/faq" component={FAQ}/>
                </Switch>
            </Router>
        );
    }
}
