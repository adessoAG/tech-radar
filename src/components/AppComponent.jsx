import React from "react";
import RadarDataService from "./RadarDataService";
import FAQ from "./FAQ";


import javaJSON from './java-radar.json';
import jsJSON from './javascript-radar.json';
import msJSON from './microsoft-radar.json';



import Tabs from "@material-ui/core/Tabs";
import Tab from "@material-ui/core/Tab";
import AppBar from "@material-ui/core/AppBar";
import { Switch, Route, Link, BrowserRouter as Router } from "react-router-dom";

import logo from "../static/img/adesso.svg";

import '../static/css/styles.scss';
import '../static/css/desctop.scss';
import '../static/css/tablet.scss';
import '../static/css/mobile.scss';


export default class AppComponent extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      isLoading: true,
      value: 0
    };
  }

  componentDidMount = () => {
    this.initializeTab();
  };

  //Based on the url, set the state ""value", which is needed for tabs, to a specific value
  initializeTab = () => {
    if (window.location.href.indexOf("/") > -1) this.setState({ value: 0 });
    if (window.location.href.indexOf("/java") > -1) this.setState({ value: 0 });
    if (window.location.href.indexOf("/microsoft") > -1) this.setState({ value: 1 });
    if (window.location.href.indexOf("/javascript") > -1) this.setState({ value: 2 });
    if (window.location.href.indexOf("/faq") > -1) this.setState({ value: 3 });
  };

  handleChange = (event, index) => {
    this.setState({
      value: index
    });
  };

  render() {
    return (
      
        <Router>
          <div id="techleiste">
            <AppBar className="tech-appbar" title="Techradar" position="static" color="default">
              
              <Tabs className="tech-tabs" centered="true" value={this.state.value} onChange={this.handleChange} textColor="primary"> 
                
                <Tab
                  className="tech-tab"
                  label={
                    <img
                      src={logo}
                      className="tech-tab-img"
                      alt="adesso's company logo" /> } />
                
                <Tab
                  className="tech-tab"
                  id="javaTab"
                  label="Java"
                  component={Link}
                  to="/java"
                  value={0} />
                
                <Tab
                  className="tech-tab"
                  id="microTab"
                  label="Microsoft"
                  component={Link}
                  to="/microsoft"
                  value={1} />
                
                <Tab
                  className="tech-tab"
                  id="jsTab"
                  label="JavaScript"
                  component={Link}
                  to="/javascript"
                  value={2} />
                
                <Tab
                  className="tech-tab"
                  id="faq"
                  label="FAQ"
                  component={Link}
                  to="/faq"
                  value={3} />
              </Tabs>
            </AppBar>
            </div>

          <Switch>
            <Route exact path="/" component = {() => (<RadarDataService data={javaJSON}/>)} />
            <Route path="/java" component={() => (<RadarDataService data={javaJSON}/>)} />
            <Route path="/microsoft" component={() => (<RadarDataService data={msJSON}/>)} />
            <Route path="/javascript" component={() => (<RadarDataService data={jsJSON}/>)} />
            <Route path="/faq" component={FAQ} />
          </Switch>

        </Router>
        
    );
  }
}


