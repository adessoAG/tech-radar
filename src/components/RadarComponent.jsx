import React from 'react'
import BlipDetailSheetComponent from './BlipDetailSheetComponent';

import Tooltip from '@material-ui/core/Tooltip';
import AppBar from '@material-ui/core/AppBar';
import Tabs from '@material-ui/core/Tabs';
import Tab from '@material-ui/core/Tab';    
import {
    createMuiTheme,
    MuiThemeProvider,
    withStyles
} from "@material-ui/core/styles";
import { Button, Icon, Slide, TextField, Container } from '@material-ui/core';
import BlipListingComponent from './BlipListingComponent';

import javaJSON from './java-radar.json'
import jsJSON from './javascript-radar.json'
import msJSON from './microsoft-radar.json'
import  MuiDownShift from 'mui-downshift';

const theme = createMuiTheme({
    overrides: {
      MuiTooltip: {
        tooltip: {
          fontSize: "1.5em"
        }
      }
    }
}), styles = {
    tooltip: {
      color: "lightblue",
      backgroundColor: "green"
    }
}, suggestions = () => {
    const list = wholeList();
    let suggestions = [];
    list.forEach((item, idx) => {
        let suggestion = {
            label: item.name,
            value: idx
        }
        suggestions.push(suggestion);
    });
    return suggestions;
}, wholeList = () => {
    let arr = [];
    for (let item of javaJSON) arr.push(item);
    for (let item of jsJSON) arr.push(item);
    for (let item of msJSON) arr.push(item);
    return arr;
}, RemoveButton = (props) => {
    return <Button className="closeButton" onClick={props.onClick}> 
        <Icon>
            clear
        </Icon>
    </Button>
}, textPosition = (i) => {
    if (i<10) {
        return 2;
    }
    if (i<100) {
        return 3.5;
    }
    return 5.5;
};


class RadarComponent extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            resolvedData: [],
            size: 0, 
            blipRadius:0, 
            centerPointX: 0,
            centerPointY: '',
            
            alphaSteps: 0,
            qOnePathDef: 0,
            qTwoPathDef: 0,
            qThreePathDef: '',
            qFourPathDef: '',
            qOneTextX: '',

            clickedBlip: null,
            showBlipDetail: false,
            transition: false,
            value: 0,
            showBlipListing: false,
            filteredBlips: [],

            filteredItems: []
        }
    }

    componentWillMount = () => {
        let arr = [];
        Promise.resolve(this.props.blips)
                .then(function(data){
                    arr = data;
                });
        setTimeout(() => {
            this.setState({ resolvedData: arr }); 
            this.initializeRadar();
        }, 100);
    }
    
    initializeRadar() {
		var kurvenPunkt = 35, sizeR = 500;
        this.qOneTextX = this.size/4;
        
        const halfRadius = sizeR/2;
        
        this.setState({
            size: sizeR,
            blipRadius: sizeR / 80,
            centerPointX: halfRadius,
            centerPointY: halfRadius,
            qOneTextX: halfRadius,
            qOnePathDef: "M 0 "+ (halfRadius-40)+" Q "+kurvenPunkt+" "+kurvenPunkt
                +" "+(halfRadius-40)+" 0",
            qTwoPathDef: "M "+(halfRadius+40)+" 0 Q "+(sizeR - kurvenPunkt)
                +" "+kurvenPunkt+" "+sizeR+" "+(halfRadius-40),
            qThreePathDef: "M "+(sizeR)+" "+(halfRadius+40)+" Q "
                +(sizeR - kurvenPunkt)+" "+(sizeR- kurvenPunkt)+" "+(halfRadius+40)+" "+sizeR,
            qFourPathDef: "M "+(halfRadius-40)+" "+sizeR
                +" Q "+kurvenPunkt+" "+(sizeR - kurvenPunkt)+" 0 "+(halfRadius+40)
        })
    }

    openInfoOfBox = (e) => {
        this.setState({
            clickedBlip: e,
            showBlipDetail: true
        })
    }

    removeBlip() {
        this.setState({showBlipDetail: false})
    }

    handleBlipListing(value) {
        setTimeout(() => {
            var quadrantCase;

            switch(value) {
                case 0:
                    this.setState( { showBlipListing: false });
                    return;
                case 1:
                    quadrantCase = 1;
                    break;
                case 2:
                    quadrantCase = 2;
                    break;
                case 3:
                    quadrantCase = 4;
                    break;
                case 4:
                    quadrantCase = 3;
                    break;
                default:
                    break;
            }
            const filteredQuadrant = this.state.resolvedData.filter(blip => blip.quadrant === quadrantCase);

            if (filteredQuadrant.length > 0) 
                this.setState({showBlipListing: true, filteredBlips: filteredQuadrant})
            else 
                this.setState({showBlipListing: false, filteredBlips: []})
        }, 100)
    }
    
    handleChange = (event, newValue) => {
        this.setState({value: newValue})
        this.handleBlipListing(newValue);
    }

    handleStateChange = changes => {
        const searchValue = changes.inputValue;
        if (typeof searchValue === 'string') {
          const filteredItems = suggestions().filter(item => item.label.toLowerCase().includes(searchValue.toLowerCase()));
          this.setState({ filteredItems });
          suggestions().find((suggestion) => {
              if (suggestion.label === searchValue) {
                  const technology = wholeList().find((e) => (e.name === searchValue));
                  this.setState({
                    showBlipDetail: true,
                    clickedBlip: technology
                  })
              }
          })
        }
    }

    render() {
        const { filteredItems, blipRadius, size, centerPointX, centerPointY,
            qOnePathDef, qTwoPathDef, qThreePathDef, qFourPathDef } = this.state;
        const showBlipDetail = this.state.showBlipDetail ? (
                <Slide in={true} direction="down" mountOnEnter unmountOnExit>
                    <div className="blip-overlay">
                        <BlipDetailSheetComponent ref={(ele) => this.blipDetail = ele } {...this.state.clickedBlip} onOutside={() => this.removeBlip()} element={<RemoveButton onClick={() => {
                            this.removeBlip();
                            }} />} 
                        />
                    </div>
                </Slide> 
            ): '',
            showBlipList = this.state.showBlipListing ? (
                <BlipListingComponent 
                    {...this.state.filteredBlips} 
                    openInfo={this.openInfoOfBox} 
                    onClick={() => this.setState({ showBlipListing: false }) } 
                />
            ): '',
            radarPanel = !this.state.showBlipListing ? (
            <div>
                <div id="autocomplete-div">
                    <Container>
                        <MuiDownShift 
                            items={filteredItems}
                            onStateChange={this.handleStateChange}
                            fullWidth={true}
                            variant="outlined"
                            getInputProps={
                                () => ({
                                    label: 'Suche Technologie . . . ',
                                    helperText: 'Gebe eine Technologie ein (z.B.: JavaFX) . . '
                                })
                            }
                        />
                    </Container>
                </div>
                <div className="radar-svg">
                <svg id="radarplot" height="90%"  viewBox="0 0 500 500">
                    <path id="q1-path" d={qOnePathDef} fill="none"/>
                        <text x="100px" className="tech-path">
                            <textPath href="#q1-path">
                                Plattformen
                            </textPath>
                        </text>

                    <path id="q2-path" d={qTwoPathDef}  fill="none"/>
                        <text x="110px" className="tech-path">
                            <textPath href="#q2-path">
                                Methoden & Techniken
                            </textPath>
                        </text>

                    <path id="q3-path" d={qThreePathDef} fill="none"/>
                        <text x="95px" className="tech-path">
                            <textPath href="#q3-path">
                                Frameworks & Sprachen
                            </textPath>
                        </text>

                    <path id="q4-path" d={qFourPathDef} fill="none"/>
                        <text x="125px" className="tech-path">
                            <textPath href="#q4-path">
                                Werkzeuge
                            </textPath>
                        </text>
                    <circle cx={centerPointX} cy={centerPointY} r={size/2}  className="outerCircleRing"/>
                    <circle cx={centerPointX} cy={centerPointY} r={size/2.5} className="middleCircleRing" />
                    <circle cx={centerPointX} cy={centerPointY} r={size/4} className="innerCircleRing"/>
                    <line className="stroke" x1={0} y1={centerPointY} x2={size} y2={size/2}/>
                    <line className="stroke" x1={centerPointX} y1={0} x2={centerPointX} y2={size}/>
                    {
                        this.state.resolvedData.map((blip, index) => {
                            return  (
                                <MuiThemeProvider theme={theme}>
                                    <Tooltip className="tooltip-blip" title={blip.name}>    
                                        <g key={index} id={blip.name} onClick={ () => this.openInfoOfBox(blip)   }  >
                                        <circle className="blipCircle" cx={blip.x}
                                            cy={blip.y}
                                            r={blipRadius}
                                        />
                                        <text className ="blipIndex" 
                                            x={blip.x - textPosition(index)} 
                                            y={blip.y + 2}
                                            fontSize={blipRadius-1}>{index+1}</text>
                                        </g>
                                    </Tooltip>
                                </MuiThemeProvider>
                            )
                        })
                    }
                </svg>
                </div>
                <div className="quadrant-buttons">
                    <button id="mobile-plattform" onClick={(e) => this.handleChange(e, 1)}>1</button>
                    <button onClick={(e) => this.handleChange(e, 2)}>2</button>
                    <button onClick={(e) => this.handleChange(e, 3)}>3</button>
                    <button onClick={(e) => this.handleChange(e, 4)}>4</button>
                </div>
            </div>
            ) : '';

       

        return <div className="radar-root">
                <div>        
                    <AppBar id="radar-appbar" centered position="static" color="default">
                        <Tabs id="quadrant-bar" className="quadrant-bar"
                            value={this.state.value}
                            onChange={this.handleChange}
                            indicatorColor="primary"
                            textColor="primary"
                            centered
                        >
                            <Tab className="quadrant-tab" id = "alleQ" label="Alle Quadranten" value={0} />
                            <Tab className="quadrant-tab" id = "plattform" label="Plattformen" value={1} />
                            <Tab className="quadrant-tab" label="Methoden/ Techniken" value={2} />
                            <Tab className="quadrant-tab" label="Frameworks & Sprachen" value={3} />
                            <Tab className="quadrant-tab" label="Werkzeuge" value={4} />
                        </Tabs>
                    </AppBar>                    
                </div>
                {radarPanel}
                
                
                {showBlipDetail}
                {showBlipList}
        </div>
    }
}

export default withStyles(styles)(RadarComponent);