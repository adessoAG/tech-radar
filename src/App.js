import React from 'react';
import './static/css/styles.scss';
import AppComponent from './components/AppComponent'

function App() {
  return (
    <AppComponent />
  );
}

export default App;
