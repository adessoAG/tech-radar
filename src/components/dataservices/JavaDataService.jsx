import RadarDataService from './RadarDataService';
import javadata from '../java-radar.json';
import React  from 'react';
import RadarComponent from '../RadarComponent';

class JavaDataService extends RadarDataService {
    constructor(props) {
        super(props);
        this.state = {
            blips: [],
            data: javadata,

            size: 500, 
            blipRadius: 0, 
            centerPointX: 0,
            centerPointY: 0,
            outerRingWidth: 0,
            middleRingWidth: 0,
            innerRingWidth: 0,
            alphaSteps: 2,
            outerRingDistance: 0,
            middleRingDistance: 0,
            innerRingDistance: 0
        }
        this.componentDidMount = this.componentDidMount.bind(this);
    }

    componentDidMount() {
        this.createBlipData();
    }

    render() {
        return <>
            <RadarComponent blips={this.state.blips} />
            </>
    }
}

export default JavaDataService;