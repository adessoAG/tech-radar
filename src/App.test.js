import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import { shallow, mount, render } from 'enzyme';
import React from 'react';
import { exportAllDeclaration } from '@babel/types';

import App from './App';
import AppComponent from './components/AppComponent';
import RadarComponent from './components/RadarComponent'
import JavaDataService from './components/dataservices/JavaDataService';


configure({ adapter: new Adapter() });

describe('<App />', () => {
    it('renders <App /> properly', () => {
      const wrapper = shallow(<App />);
      expect(wrapper).toMatchSnapshot();
    });
});  

describe('<AppComponent />', () => {
    it('renders <AppComponent /> properly', () => {
        const wrapper = shallow(<AppComponent />);
        expect(wrapper).toMatchSnapshot();
    })
})

describe('<JavaDataService /> \n', () => {
    it('renders <JavaDataService /> component properly', () => {
        const wrapper = shallow(<JavaDataService />);
        expect(wrapper).toMatchSnapshot();
    })
    it('find Java in Component', () => {
        const wrapper = shallow(<JavaDataService />);
        const blips = Object.values(wrapper.state('blips'));
        for (let item of blips) {
            if (item.name === "Java") {
                console.log(item);
                break;
            }
        }
        const buttons = wrapper.filter('g');
        const java = buttons.find({ title: "Java"})
        console.log(buttons.debug());
    })
})

