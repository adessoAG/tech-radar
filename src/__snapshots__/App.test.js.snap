// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> renders <App /> properly 1`] = `ShallowWrapper {}`;

exports[`<AppComponent /> renders <AppComponent /> properly 1`] = `ShallowWrapper {}`;

exports[`<JavaDataService /> 
 renders <JavaDataService /> component properly 1`] = `ShallowWrapper {}`;
